File: Infrastructure/src/ServiceStack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.DynamoDB;
using Amazon.CDK.AWS.IAM;
using Amazon.CDK.AWS.IoT;
using Amazon.CDK.AWS.Lambda;
using Amazon.CDK.AWS.Lambda.EventSources;
using Constructs;
using static Amazon.CDK.AWS.IoT.CfnTopicRule;

namespace Infrastructure
{
  public class DevicesDistanceTrackerInfrastructure : Stack
  {
    internal DevicesDistanceTrackerInfrastructure(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
    {
      var LambdaFunctionName = "DistanceTrackerFunction";
      var DynamoDbTable = new Table(this, "DevicesLocationTable", new TableProps
      {
        TableName = "DevicesLocation",
        PartitionKey = new Attribute { Name = "macAddress", Type = AttributeType.STRING },
        BillingMode = BillingMode.PROVISIONED,
        ReadCapacity = 5,
        WriteCapacity = 5,
        RemovalPolicy = RemovalPolicy.RETAIN,
        Stream = StreamViewType.NEW_AND_OLD_IMAGES,
      });

      foreach (var tag in this.GetDefaultTags())
      {
        Amazon.CDK.Tags.Of(DynamoDbTable).Add(tag.Key, tag.Value);
      }

      var IoTRuleExecutionRole = new Role(this, "IotRuleExecutionRole", new RoleProps
      {
        AssumedBy = new ServicePrincipal("iot.amazonaws.com"),
        RoleName = "IotRuleExecutionRole",
      });
      IoTRuleExecutionRole.AddToPolicy(new PolicyStatement(this.GetDynamoDBPutPolicyDocument(DynamoDbTable.TableName)));
      foreach (var tag in this.GetDefaultTags())
      {
        Amazon.CDK.Tags.Of(IoTRuleExecutionRole).Add(tag.Key, tag.Value);
      }

      var LambdaFunctionExecutionRole = new Role(this, "LambdaFunctionExecutionRole", new RoleProps
      {
        AssumedBy = new ServicePrincipal("lambda.amazonaws.com"),
        RoleName = "LambdaFunctionExecutionRole",
      });
      foreach (var Policy in this.GetLambdaFunctionPolicyDocument(LambdaFunctionName, DynamoDbTable.TableName, DynamoDbTable.TableStreamArn))
      {
        LambdaFunctionExecutionRole.AddToPolicy(new PolicyStatement(Policy));
      }
      foreach (var tag in this.GetDefaultTags())
      {
        Amazon.CDK.Tags.Of(LambdaFunctionExecutionRole).Add(tag.Key, tag.Value);
      }

      var IotRule = new CfnTopicRule(this, "DevicesLocationRule", new CfnTopicRuleProps
      {
        TopicRulePayload = new TopicRulePayloadProperty
        {
          Actions = new[] { new ActionProperty {
            DynamoDBv2 = new DynamoDBv2ActionProperty {
                PutItem = new PutItemInputProperty {
                    TableName = DynamoDbTable.TableName
                },
                RoleArn = IoTRuleExecutionRole.RoleArn
            }} },
          Sql = "SELECT CASE isUndefined(vehicleId) WHEN true then handheldId else vehicleId end AS macAddress, CASE isUndefined(vehicleId) WHEN true then 'handheld' else 'vehicle' end AS type, * FROM 'v1/gps/+/#'",
          AwsIotSqlVersion = "2016-03-23",
          Description = "IoT Rule to receice Data from Devices and send them to DynamoDB table.",
        },
        RuleName = "DevicesLocationRule",
        Tags = this.GetDefaultTags()
      });

      var DistanceTrackerLambdaFunction = new Function(this, "DistanceTrackerFunction", new FunctionProps
      {
        Runtime = Runtime.DOTNET_6,
        MemorySize = 128,
        Architecture = Architecture.ARM_64,
        Handler = "DistanceTrackerFunction::DistanceTrackerFunction.Function::FunctionHandler",
        Code = Code.FromAsset("deploy/DistanceTrackerFunction"),
        Role = LambdaFunctionExecutionRole,
        FunctionName = LambdaFunctionName,
        Timeout = Duration.Seconds(10),
      });
      foreach (var tag in this.GetDefaultTags())
      {
        Amazon.CDK.Tags.Of(DistanceTrackerLambdaFunction).Add(tag.Key, tag.Value);
      }
      DistanceTrackerLambdaFunction.AddEventSource(new DynamoEventSource(DynamoDbTable, new DynamoEventSourceProps { StartingPosition = StartingPosition.LATEST, BatchSize = 100, MaxBatchingWindow = Duration.Seconds(15), RetryAttempts = 0 }));
    }
    private CfnTag[] GetDefaultTags()
    {
      return new CfnTag[] { new CfnTag { Key = "Project", Value = "DevicesDistanceTracker" }, new CfnTag { Key = "Version", Value = "1.0.0" } };
    }

    private PolicyStatementProps GetDynamoDBPutPolicyDocument(string TableName)
    {
      return new PolicyStatementProps
      {
        Effect = Effect.ALLOW,
        Actions = new[] { "dynamodb:PutItem" },
        Resources = new[]
                {
                    Arn.Format(new ArnComponents
                    {
                        Service = "dynamodb",
                        Resource = "table",
                        ResourceName = TableName
                    },this)
                }
      };
    }
    private PolicyStatementProps[] GetLambdaFunctionPolicyDocument(string lambdaName, string devicesTableName, string streamArn)
    {
      return new PolicyStatementProps[]
      {
       new PolicyStatementProps {
          Effect = Effect.ALLOW,
          Actions = new[] { "logs:CreateLogGroup" },
          Resources = new[]
                  {
                      Arn.Format(new ArnComponents
                      {
                          Service = "logs",
                          Resource = "*"
                      },this)
                  }
        },
        new PolicyStatementProps {
          Effect = Effect.ALLOW,
          Actions = new[] { "logs:CreateLogStream", "logs:PutLogEvents" },
          Resources = new[]
                  {
                      Arn.Format(new ArnComponents
                      {
                          Service = "logs",
                          Resource = "log-group",
                          ResourceName = $"/aws/lambda/{lambdaName}",
                          ArnFormat = ArnFormat.COLON_RESOURCE_NAME,
                      },this),
                      Arn.Format(new ArnComponents
                      {
                          Service = "logs",
                          Resource = "log-group",
                          ResourceName = $"/aws/lambda/{lambdaName}:*",
                          ArnFormat = ArnFormat.COLON_RESOURCE_NAME,
                      },this)
                  }
        },
        new PolicyStatementProps {
          Effect = Effect.ALLOW,
          Actions = new[] { "dynamodb:GetItem" },
          Resources = new[]
                  {
                      Arn.Format(new ArnComponents
                      {
                          Service = "dynamodb",
                          Resource = "table",
                          ResourceName = $"{devicesTableName}"
                      },this),
                      Arn.Format(new ArnComponents
                      {
                          Service = "dynamodb",
                          Resource = "table",
                          ResourceName = "DevicePairs"
                      },this)
                  },
        },
        new PolicyStatementProps {
          Effect = Effect.ALLOW,
          Actions = new[] { "sns:Publish" },
          Resources = new[]
                  {
                      Arn.Format(new ArnComponents
                      {
                          Service = "sns",
                          Resource = "PostNL_Topic",
                      },this)
                  }
        },
        new PolicyStatementProps {
          Effect = Effect.ALLOW,
          Actions = new[] { "dynamodb:DescribeStream", "dynamodb:GetRecords", "dynamodb:GetShardIterator" },
          Resources = new[]
                  {
                      streamArn
                  }
        },
        new PolicyStatementProps {
          Effect = Effect.ALLOW,
          Actions = new[] { "dynamodb:ListStreams" },
          Resources = new[]
                  {
                      "*"
                  }
        },
      };
    }
  }
}

File: Infrastructure/src/lib/serviceStack.ts
import * as cdk from 'aws-cdk-lib';
import { Arn, ArnFormat, Duration, RemovalPolicy } from 'aws-cdk-lib';
import { AttributeType, BillingMode, StreamViewType, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Effect, PolicyDocument, PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { CfnTopicRule } from 'aws-cdk-lib/aws-iot';
import { Architecture, CfnEventSourceMapping, Code, Function, Runtime, StartingPosition, Tracing } from 'aws-cdk-lib/aws-lambda';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Construct } from 'constructs';


export class ServiceStack extends cdk.Stack {
  private devicesDynamoDbTable: Table;
  private iotExecutionRole: Role;
  private lambdaExecutionRole: Role;
  private iotRule: CfnTopicRule;
  private distanceTrackerLambdaFunction: Function;
  private dynamoDbEventSource: CfnEventSourceMapping;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const lambdaFunctionName = "DistanceTrackerFunction";

    this.devicesDynamoDbTable =  new Table(this,"DevicesLocationTable", {
      tableName: "DevicesLocation",
      partitionKey: {name: "macAddress", type: AttributeType.STRING},
      billingMode: BillingMode.PROVISIONED,
      readCapacity: 5,
      writeCapacity: 5,
      removalPolicy: RemovalPolicy.RETAIN,
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
      } );

    this.iotExecutionRole = new Role(this, "IotRuleExecutionRole", {
      assumedBy: new ServicePrincipal("iot.amazonaws.com"),
      roleName: "IotRuleExecutionRole",
      inlinePolicies: {
        "IoTRuleExecutionRoleDynamoDbPolicy": new PolicyDocument({
          statements:  [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["dynamodb:PutItem"],
              resources: [
                Arn.format({
                  service: "dynamodb",
                  resource: "table",
                  resourceName: this.devicesDynamoDbTable.tableName,
                }, this)
              ]

            }),
          ]
        }),
      }
    });

    this.lambdaExecutionRole = new Role(this, "LambdaFunctionExecutionRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      roleName: "LambdaFunctionExecutionRole",
      inlinePolicies: {
        "IoTRuleExecutionRolePolicy": new PolicyDocument({
          statements:  [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["logs:CreateLogGroup"],
              resources: [
                Arn.format({
                  service: "logs",
                  resource: "*",
                }, this)
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["logs:CreateLogStream", "logs:PutLogEvents"],
              resources: [
                Arn.format({
                  service: "logs",
                  resource: "log-group",
                  resourceName: `/aws/lambda/${lambdaFunctionName}`,
                  arnFormat: ArnFormat.COLON_RESOURCE_NAME,
                }, this),
                 Arn.format({
                  service: "logs",
                  resource: "log-group",
                  resourceName: `/aws/lambda/${lambdaFunctionName}:*`,
                  arnFormat: ArnFormat.COLON_RESOURCE_NAME,
                }, this)
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["dynamodb:GetItem"],
              resources: [
                Arn.format({
                  service: "dynamodb",
                  resource: "table",
                  resourceName: this.devicesDynamoDbTable.tableName,
                }, this),
                Arn.format({
                  service: "dynamodb",
                  resource: "table",
                  resourceName: "DevicePairs",
                }, this)
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["sns:Publish"],
              resources: [
                Arn.format({
                  service: "sns",
                  resource: "PostNL_Topic",
                }, this)
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["dynamodb:DescribeStream", "dynamodb:GetRecords", "dynamodb:GetShardIterator"],
              resources: [
               this.devicesDynamoDbTable.tableStreamArn as string,
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["dynamodb:ListStreams"],
              resources: [
               "*"
              ]
            })
          ]
        }),
      }
    });

    this.iotRule = new CfnTopicRule(this, "DevicesLocationRule", {
      ruleName: "DevicesLocationRule",
      topicRulePayload: {
        actions: [
          {
            dynamoDBv2: {
              putItem: {
                tableName: this.devicesDynamoDbTable.tableName,
              },
              roleArn: this.iotExecutionRole.roleArn,
            }
          }
        ],
        sql: "SELECT CASE isUndefined(vehicleId) WHEN true then handheldId else vehicleId end AS macAddress, CASE isUndefined(vehicleId) WHEN true then 'handheld' else 'vehicle' end AS type, * FROM 'v1/gps/+/#'",
        awsIotSqlVersion: "2016-03-23",
      }
    });

    this.distanceTrackerLambdaFunction = new Function(this, "DistanceTrackerFunction", {
      runtime: Runtime.DOTNET_6,
      memorySize: 512,
      architecture: Architecture.ARM_64,
      handler: "DistanceTrackerFunction::DistanceTrackerFunction.Function::FunctionHandler",
      code: Code.fromAsset("deploy/DistanceTrackerFunction"),
      role: this.lambdaExecutionRole,
      functionName: lambdaFunctionName,
      timeout: Duration.seconds(15),
      tracing: Tracing.ACTIVE
    })


    this.dynamoDbEventSource = new CfnEventSourceMapping(this, "DynamoDbEventSource", {
      startingPosition: StartingPosition.LATEST,
      batchSize: 100,
      maximumBatchingWindowInSeconds: 10,
      maximumRetryAttempts: 0,
      functionName: this.distanceTrackerLambdaFunction.functionName,
      eventSourceArn: this.devicesDynamoDbTable.tableStreamArn
    });
    
    this.dynamoDbEventSource.addPropertyOverride("FilterCriteria", {
      Filters: [
        {
          Pattern:
            JSON.stringify({
              dynamodb: {
                NewImage: {
                  type: {S: ['vehicle']},
                },
              },
              eventName: ['MODIFY'],
            }),
        },
      ],
    });    
  }
}
